### PR TITLE
API development

### PR DESCRIPTION
**Pull request by zsz12251665**
I was working on the APIs. I have found some problems and make some difference and this PR is here to log it.

### CODE REVIEW

**Comment by zsz12251665** [comment]
Currently the API response difference is as followed (UPDATED@ 2021-12-25 16:16):

- For all APIs, there is a try/catch block covering the function which returns HTTP 500 with error message of the server
- For all APIs, there is a handler validating the HTTP request method and throws HTTP 405 if there is no function bind to the method
- For those APIs with empty response bodies, the POST request will earn a HTTP 204 instead of HTTP 200
- For `/api/user` & `/api/notifications` & `/api/books/:isbn/comments`, the HTTP 400 is changed to HTTP 500, which is the difference above
- For `/api/user/:userId`, wrong password will result in HTTP 403 instead of HTTP 400
- For `/api/user/:userId/borrow`, it will return HTTP 404 when the user with specified ID does not exists
- For `/api/notifications/:id`, it will return HTTP 404 when the notification with specified ID does not exists
- For `/api/books/:isbn/comments`, comment adding has been implemented

These changes could be found in code by searching `//!`

**Comment by zsz12251665** [comment]
There are also some unhandled problems (UPDATE@ 2021-12-25 17:39):

- ***Unsolved*** For `/api/login`, the authentication token is a TODO. ~~What is the payload of the token? Is it returned in response body or written to cookies directly?~~ Cookie is preferred, but the token is still not implemented yet
- *(Solved)* ~~For `/api/user/`, password is a rather sensitive field which the database stores only the hash string and salt string instead of the plain string. Is it required?~~ password field in `UserInfo` has been removed
- *(Solved)* ~~For transaction fines, it cannot be calculated until the book is returned. Currently it returns `NaN` when unreturned. Is it fine?~~ 0 is used instead
- *(Solved)* ~~For borrowing and returning process, we need to replace the book ISBN with the copy ID since the transaction tracks the copy instead of the book directly.~~ Fixed in PR #4
- *(Solved)* ~~For pagination requests, `pageCount` in response is currently the number of all suitable entries instead of the number of pages. Is that fine? Or shall we need a default page limit when the search string does not provided limit nor offset? (According to the SQL, current default values are 0 for offset and &infin; for limit)~~ `pageCount` renamed to `totalt`
- *(Solved)* ~~Fuzzy search (`filter`) for book searching has not been implemented yet~~ Implemented with `MATCH()` in fulltext index in SQL
- ***Unsolved*** The `available` field in BookInfo is not fulfilled (currently placeheld by NaN). The usage of the field requires discussion since the borrow is copy-based now
- *(Solved)* ~~For `/api/books/:isbn/comments`, `isbn` field in the body of the post request is redundant since it can be inferred from URL, and the data type of `userId` in `CommentAddRequest` is also wrong (`string` expected, got `number` instead)~~ Fixed
- ***Unsolved*** For the `userId` included in the search strings, it is recommended to extract it from the token instead

These problems could be found in code by searching `//?`

**Comment by yioneko** [comment]
For token, both solutions are feasible, but I prefer cookies because it leads to fewer modifications to the current codebase, and we don't need to update some api interfaces to include the token field and implement token storage and retrieval logic in frontend.

~~Password can be either encrypted in client side first (need agreed salt or algorithm from server) or sent directly to server and let server deal with it. Technically the former way is more preferable but involves more works to do (for example, synchronize the salt value), it just depends on the time.~~ (EDIT: Problem is misunderstood)

Instead of NaN, why not return 0 ? The semantics of NaN is more like “an error happens during numeric operation", which I don't think applies to this situation (just meaning that we haven't calculated it yet).

**Comment by zsz12251665** [comment]
Well, currently the salting and hashing is handled in [database-coursedesign-model](https://github.com/zsz12251665/database-course-design-model/blob/master/src/DAO/User.ts#L34). The problem is that by `common/interface.ts`, the password is required when getting users' information. But it should not and cannot be offered.

**Comment by yioneko** [comment]
> Well, currently the salting and hashing is handled in [database-coursedesign-model](https://github.com/zsz12251665/database-course-design-model/blob/master/src/DAO/User.ts?rgh-link-date=2021-12-24T04%3A34%3A57Z#L34). The problem is that by `common/interface.ts`, the password is required when getting users' information. But it should not and cannot be offered.

Here?
https://github.com/yioneko/database-course-design-web/blob/077609c763fb4e12eda3122cf01a97279e1a7f80/common/interface.ts#L75-L77

Yes the type is wrong, password should be excluded.

**Comment by zsz12251665** [comment]
Done

File: common/interface.ts [hunk 1]
```diff
@@ -1,7 +1,6 @@
 export interface UserInfo {
   userId: string;
   name: string;
-  password: string;
```
Comment (yioneko) [inline common/interface.ts:4]: Here should be changed either:
https://github.com/yioneko/database-course-design-web/blob/077609c763fb4e12eda3122cf01a97279e1a7f80/common/interface.ts#L48

Comment (zsz12251665) [reply common/interface.ts:4]: Done

**Comment by zsz12251665** [comment]
https://github.com/yioneko/database-course-design-web/blob/077609c763fb4e12eda3122cf01a97279e1a7f80/common/interface.ts#L100

The borrowing process seems different from what it should be like. The transaction tracks Copy instead of Book. The front end should submit copy ID instead of book ISBN.

**Comment by yioneko** [comment]
Reasonable. Originally I think backend should select a copy randomly🤣. I'll fix it in another branch for frontend.

**Comment by zsz12251665** [comment]
https://github.com/yioneko/database-course-design-web/blob/4e5e1627f11bd9acd001d8bdfab744b28ff73b04/common/interface.ts#L22

For notifications, the database does not care who is the sender. I suppose that the user just need to know it is from the administrator. (Damn I should have thought of these interface problems in PR #1😓)

**Comment by zsz12251665** [comment]
https://github.com/yioneko/database-course-design-web/blob/4e5e1627f11bd9acd001d8bdfab744b28ff73b04/mocks/handlers.ts#L126-L134

I failed understand what this interface is doing.

According to the generic, `req.body` is undefined, so it affects nothing to assign it to the notification object. Is there an `isRead` field in `req.body` actually?

**Comment by yioneko** [comment]
Sorry for mistake here😅, `req.body` does contain an `isRead` field.

**Comment by yioneko** [comment]
> Reasonable. Originally I think backend should select a copy randomly🤣. I'll fix it in another branch for frontend.

Fixed at #4

**Comment by yioneko** [comment]
> https://github.com/yioneko/database-course-design-web/blob/4e5e1627f11bd9acd001d8bdfab744b28ff73b04/common/interface.ts#L22
> 
> For notifications, the database does not care who is the sender. I suppose that the user just need to know it is from the administrator. (Damn I should have thought of these interface problems in PR #1😓)

Fixed at #6

**Comment by zsz12251665** [comment]
https://github.com/yioneko/database-course-design-web/blob/e945f9940d26f62693b3978e6d6c821156557c7d/common/interface.ts#L134-L139

`isbn` field in the body of the post request is redundant since it can be inferred from URL. The data type of `userId` in is also wrong (`string` expected, got `number` instead)

**Comment by zsz12251665** [comment]
(Check the top 2 comments for more latest important modifications and problems 🙂

**Comment by yioneko** [comment]
Does it matter that I push commits directly to this branch? All the issues are not that hard to fix (except implementing token, I'm still considering solutions for it).

**Comment by zsz12251665** [comment]
I think it's fine

**Comment by zsz12251665** [comment]
I think this branch is ready to be merged. We may leave the token to be fixed on another branch.

**Pull request merged**